New challenge - MD5 hashing

RemakingEden opened the pull request:
This is a challenge for the issue https://github.com/commjoen/wrongsecrets/issues/200. Using a weak MD5 hash that is easily crackable.

This is just the first draft but I thought I would get it out early so the community can have a look and make sure im on the right track.

My current questions on this challenge are:
- It currently seems quite easy, but it will require some googling if you havent heard about MD5 hashes. Is this good enough or should we add complexity?
- I have the hash listed in the challenge text, is it worth adding this into the container somehow? I was going to do that but then it felt a bit like I was adding complexity with no real gain for the user. The only gain I can think of is letting the user figure out how to search for MD5 hashes.

To do:

 - [x] Encode the answer in the challenge so a user cannot simply look at the source code
 - [x] Add working tests
 - [x] Proof read language and ensure it makes sense/is succinct
- [x] Check if Trufflehog or git-secrets will find MD5 hashes

bendehaan commented on src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge18.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public boolean answerCorrect(String answer) {
+        return "hunter2".equals(answer);
```
Would it be an idea to compare it to a hash of the input? This is "greppable". Of course that would mean revising the spoiler as well - probably that could work along the lines of challenge 15 (w/ encryption)

RemakingEden replied:
I compared to the hash in the end, As for the spoiler this wasnt possible so ive ended up using bse64 to obfuscate. I figured this is enough as if the user really want to cheat they can just use the spoiler.

bendehaan replied:
This solution looks good to me!

bendehaan reviewed (commented):
Hi Joss, thanks for your contribution, much appreciated! 

I think it's nice to have a hashcat challenge in there, and this is probably complex enough. I'm fine with keeping the hash in the description and eventually changing it if feedback shows us this is too easy.

RemakingEden commented:
> Hi Joss, thanks for your contribution, much appreciated!
> 
> I think it's nice to have a hashcat challenge in there, and this is probably complex enough. I'm fine with keeping the hash in the description and eventually changing it if feedback shows us this is too easy.

Thanks for reviewing Ben. I added a little more to the challenge by putting both MD5 and SHA1 into the challenge now. What do you think? Ive also added tests and obfuscated the answer in the source, its still easy to find if you decode the base64 but its not right there in your face when you look at the code.

RemakingEden commented:
This import "org.apache.commons.codec.digest.DigestUtils" has been marked as illegal by the checkstyle. Is there a reason for this? Do I need to find a workaround?

bendehaan commented:
> This import "org.apache.commons.codec.digest.DigestUtils" has been marked as illegal by the checkstyle. Is there a reason for this? Do I need to find a workaround?

This is the part of the checkstyle config it matches, but it doesn't seem like it should:
```xml
<property name="illegalPkgs" value="^sun.*, ^org\.apache\.commons\.(?!compress|dbcp2|lang|lang3|logging|io|pool2).*, ^org\.flywaydb\.core\.internal.*, ^org\.testcontainers\.shaded.*"/>
```

I think there's a missing `\` in the regex right after the `(compress|....|pool2)`, you could change it to:
```xml
<property name="illegalPkgs" value="^sun.*, ^org\.apache\.commons\.(?!compress|dbcp2|lang|lang3|logging|io|pool2)\.*, ^org\.flywaydb\.core\.internal.*, ^org\.testcontainers\.shaded.*"/>
```

Then:
```
./mvnw --no-transfer-progress checkstyle:check

[INFO] Scanning for projects...
[INFO] 
[INFO] -----------------------< org.owasp:wrongsecrets >-----------------------
[INFO] Building OWASP WrongSecrets 1.4.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-checkstyle-plugin:3.1.2:check (default-cli) @ wrongsecrets ---
[INFO] Starting audit...
Audit done.
[INFO] You have 0 Checkstyle violations.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.389 s
[INFO] Finished at: 2022-05-27T10:55:35+02:00
[INFO] ------------------------------------------------------------------------
```


@commjoen does that also sound okay with you or is this deliberate? And does this need to be in `pom.xml`?


```xml
    <properties>
   ...
        <commons-codec.version>1.15</commons-codec.version>
    </properties>
```

```xml
    <dependencies>
    ...
        <dependency>
            <groupId>commons-codec</groupId>
            <artifactId>commons-codec</artifactId>
            <version>${commons-codec.version}</version>
        </dependency>
    </dependencies>
```

RemakingEden commented:
@bendehaan Yes I see the apparent error in the checkstyle. It seems like a backslash is needed to escape the wildcard "." but then I have added another wildcard "." as I think the intent was to include anything after.

i have also edited the hint a little to make it more readable and changed the Challenge 18 code to always use the calculateHash method when creating hashes. Before it was sometimes using this method and sometimes just doing it in its raw form which was not good.

@commjoen let me know about the pom file!

commjoen commented on src/main/resources/explanations/challenge18_reason.adoc:
```diff
@@ -0,0 +1,7 @@
+
+MD5 and SHA1 hash are no longer considered safe to store passwords on their own. Speed is what makes MD5 and SHA1 hashes so useful, but it is also their downfall. It only takes a few minutes to hash thousands of passwords; this also means that it only takes minutes to hash thousands of common passwords and use these hashes to compare against a hash that has been obtained.
+
+Companies try different techniques to harden MD5 and SHA1 hashes, such as "salting" them. This is the process of adding additional characters to the password that only the person/company that should be decrypting knows.
```
```suggestion
Companies try different techniques to harden MD5 and SHA1 hashes, such as "salting" them. This is the process of adding additional characters to the password that only the person/company that should be decrypting knows. Unfortunately this is not enough either with the raise of GPU and ASIC based computations. Therefore, companies using these techniques can better migrate to [Argon2](https://en.wikipedia.org/wiki/Argon2) or [Balloon hashing](https://en.wikipedia.org/wiki/Balloon_hashing).
```

commjoen commented on src/main/resources/explanations/challenge18_reason.adoc:
```diff
@@ -0,0 +1,7 @@
+
+Companies try different techniques to harden MD5 and SHA1 hashes, such as "salting" them. This is the process of adding additional characters to the password that only the person/company that should be decrypting knows.
+
+As a user you often have no choice in how your passwords are stored; the only thing you can do in this case is try to make your password longer and more complex. A password hashed with MD5 and SHA1 that is 7 characters long with upper and lowercase characters will take roughly 25 seconds to brute force, whereas one with 18 characters will take roughly 6 trillion years. That would take some real p ...
```
```suggestion
As a user you often have no choice in how your passwords are stored; the only thing you can do in this case is try to make your password longer and more complex. A password SHA1 that is 7 characters long with upper and lowercase characters will take roughly a minute or 5 to brute force on a proper GPU, whereas one with 25 characters will take much longer. That would take some real patience.
```

commjoen commented:
https://github.com/commjoen/wrongsecrets/pull/298#issuecomment-1139423892 

> @bendehaan Yes I see the apparent error in the checkstyle. It seems like a backslash is needed to escape the wildcard "." but then I have added another wildcard "." as I think the intent was to include anything after.
> 
> i have also edited the hint a little to make it more readable and changed the Challenge 18 code to always use the calculateHash method when creating hashes. Before it was sometimes using this method and sometimes just doing it in its raw form which was not good.
> 
> @commjoen let me know about the pom file!

The `org.apache.commons.codec.digest.DigestUtils` requires you to import it i guess &update the pom file?
What if you did the following for calculating the hash instead?

```java
MessageDigest md = MessageDigest.getInstance("SHA-1");
byte[] inputBytes;
inputBytes = input.getBytes("UTF-8");
return md.digest(inputBytes);
```

commjoen commented:
PSSST, i might not have said itin thisPR yet: but you are doing an awesome job!!! REally love the new challenge already & am looking forward to the point where we can put it in our beta-test env!

RemakingEden commented:
> ```java
> MessageDigest md = MessageDigest.getInstance("SHA-1");
> byte[] inputBytes;
> inputBytes = input.getBytes("UTF-8");
> return md.digest(inputBytes);
> ```

Ive just looked into this and it seems that creating the hash this way is a lot more complicated and involves importing other packages, your example @commjoen only gives the bytes and then converting it back to a string seems unnecessarily complicated and hard to read. Is there a big issue with us adding this to the POM and using it this way?

RemakingEden commented:
Okay @commjoen & @bendehaan Ive added it to the pom.xml, im very new to Maven so please can you double check i did it correctly? Apart from that I guess its ready for its final review!
